File: backend/src/core/executor.py
"""Single step executor - calls Claude Agent SDK."""
import asyncio
import os
from dataclasses import dataclass
from typing import Optional

import structlog

logger = structlog.get_logger()

# Clean CLAUDECODE env var at module load to prevent nested session conflicts
os.environ.pop("CLAUDECODE", None)


@dataclass
class StepResult:
    output: str = ""
    tokens_used: int = 0
    model_used: str = ""
    failed: bool = False
    error: Optional[str] = None


async def execute_step(step_config: dict, context: dict, merged_config: dict) -> StepResult:
    """Execute a single step by calling Claude Agent SDK.

    Args:
        step_config: Step configuration {name, prompt, tools, mcp, rules, model, max_turns}
        context: Current execution context with inputs and previous step outputs
        merged_config: Merged global + workflow config {system_prompt, allowed_tools, mcp_servers}
    """
    try:
        os.environ.pop("CLAUDECODE", None)

        from claude_code_sdk import (
            AssistantMessage,
            ClaudeCodeOptions,
            ResultMessage,
            TextBlock,
            query,
        )

        # Render prompt with context
        from src.core.template import render_template

        prompt = render_template(step_config.get("prompt", ""), context)

        # Build system prompt
        system_prompt = merged_config.get("system_prompt", "")
        step_rules = step_config.get("rules", {})
        if step_rules and step_rules.get("system_prompt"):
            system_prompt = f"{system_prompt}\n\n{step_rules['system_prompt']}"

        # Build options
        options_kwargs = {}
        if system_prompt:
            options_kwargs["system_prompt"] = system_prompt

        allowed_tools = merged_config.get("allowed_tools")
        if allowed_tools:
            options_kwargs["allowed_tools"] = allowed_tools

        max_turns = step_config.get("max_turns") or 30
        options_kwargs["max_turns"] = max_turns

        # MCP servers
        mcp_servers = merged_config.get("mcp_servers")
        if mcp_servers:
            options_kwargs["mcp_servers"] = mcp_servers

        options = ClaudeCodeOptions(**options_kwargs)

        # Apply timeout if configured
        max_duration = merged_config.get("max_duration")

        # Call Agent SDK
        output_parts = []
        total_tokens = 0
        model_used = ""

        async def run_query():
            nonlocal total_tokens, model_used
            async for message in query(prompt=prompt, options=options):
                # Handle AssistantMessage: extract text from content blocks
                if isinstance(message, AssistantMessage):
                    if hasattr(message, "model") and message.model:
                        model_used = message.model
                    if hasattr(message, "content") and isinstance(message.content, list):
                        for block in message.content:
                            if isinstance(block, TextBlock):
                                output_parts.append(block.text)

                # Handle ResultMessage: extract usage stats
                elif isinstance(message, ResultMessage):
                    if hasattr(message, "result") and message.result:
                        output_parts.append(message.result)
                    if hasattr(message, "usage") and message.usage:
                        usage = message.usage
                        if isinstance(usage, dict):
                            total_tokens += usage.get("input_tokens", 0) + usage.get("output_tokens", 0)

        if max_duration:
            await asyncio.wait_for(run_query(), timeout=max_duration)
        else:
            await run_query()

        output_text = "\n".join(output_parts) if output_parts else ""

        return StepResult(
            output=output_text,
            tokens_used=total_tokens,
            model_used=model_used,
        )

    except asyncio.TimeoutError:
        return StepResult(failed=True, error="Step execution timed out")
    except Exception as e:
        logger.error("step_execution_failed", error=str(e))
        return StepResult(failed=True, error=str(e))

File: electron-app/src/main/core/executor.ts
/**
 * 单步执行器
 *
 * 调用 Claude Agent SDK 执行单个步骤
 *
 * @author zhourui(V33215020)
 * @since 2026/03/11
 */

import { query } from '@anthropic-ai/claude-code';
import log from 'electron-log';
import type { MergedConfig, StepResult } from '../store/models';

/**
 * 从消息内容中提取文本
 *
 * @param content 消息内容
 * @returns 提取的文本
 */
function extractText(content: unknown): string {
  if (typeof content === 'string') {
    return content;
  }

  if (Array.isArray(content)) {
    return content
      .filter((block): block is { type: string; text: string } =>
        typeof block === 'object' &&
        block !== null &&
        'type' in block &&
        block.type === 'text' &&
        'text' in block
      )
      .map(block => block.text)
      .join('');
  }

  return '';
}

/**
 * 执行单个步骤
 *
 * @param prompt 渲染后的提示词
 * @param config 合并后的配置
 * @param onProgress 进度回调
 * @returns 步骤执行结果
 */
export async function executeStep(
  prompt: string,
  config: MergedConfig,
  onProgress?: (text: string) => void
): Promise<StepResult> {
  let outputText = '';
  let tokensUsed = 0;

  log.debug(`Executing step with prompt length: ${prompt.length}`);

  try {
    const q = query({
      prompt,
      options: {
        model: config.model || 'claude-sonnet-4-20250514',
        customSystemPrompt: config.systemPrompt,
        allowedTools: config.allowedTools,
        mcpServers: config.mcpServers as Record<string, import('@anthropic-ai/claude-code').McpServerConfig> | undefined,
        maxTurns: config.maxTurns || 30,
        permissionMode: 'acceptEdits',
        cwd: process.cwd()
      }
    });

    for await (const msg of q) {
      if (msg.type === 'assistant') {
        const text = extractText(msg.message?.content);
        if (text) {
          outputText += text;
          onProgress?.(text);
        }
      }

      if (msg.type === 'result') {
        const usage = (msg as { usage?: { total_tokens?: number } }).usage;
        tokensUsed = usage?.total_tokens || 0;
        log.debug(`Step completed with ${tokensUsed} tokens`);
      }
    }

    return {
      success: true,
      outputText,
      tokensUsed
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    log.error('Step execution failed:', errorMessage);

    return {
      success: false,
      outputText,
      tokensUsed,
      errorMessage
    };
  }
}

/**
 * 带超时的步骤执行
 *
 * @param prompt 渲染后的提示词
 * @param config 合并后的配置
 * @param timeoutMs 超时时间（毫秒）
 * @param onProgress 进度回调
 * @returns 步骤执行结果
 */
export async function executeStepWithTimeout(
  prompt: string,
  config: MergedConfig,
  timeoutMs: number,
  onProgress?: (text: string) => void
): Promise<StepResult> {
  const timeoutPromise = new Promise<StepResult>((_, reject) => {
    setTimeout(() => {
      reject(new Error(`Step execution timed out after ${timeoutMs}ms`));
    }, timeoutMs);
  });

  const executionPromise = executeStep(prompt, config, onProgress);

  try {
    return await Promise.race([executionPromise, timeoutPromise]);
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);
    return {
      success: false,
      outputText: '',
      tokensUsed: 0,
      errorMessage
    };
  }
}
